File: R/fetch_europepmc.R
# Europe PMC REST search.
#
# One robust endpoint that also folds in medRxiv preprints and Cochrane review
# metadata. resultType=core returns abstracts. We page with cursorMark. Where
# Europe PMC tags an item as a randomized controlled trial we record that in
# `design`, which lets you filter out the reviews/protocols the broad query
# also returns.

.EPMC_BASE <- "https://www.ebi.ac.uk/europepmc/webservices/rest/search"

fetch_europepmc <- function(cfg, start_date, end_date) {
  # Malaria scoped to title/abstract/keyword/MeSH (not incidental full-text).
  # RCT signal = MEDLINE-curated RCT publication type OR a preprint (SRC:PPR)
  # whose title carries a randomi* token. Europe PMC's unique value is preprints,
  # which have no pub-type tag; the wildcard MUST be unquoted (a quoted "randomi*"
  # is matched literally and returns nothing). PubMed carries recall for published
  # papers; the eligibility screen supplies precision.
  query <- sprintf(paste0(
    '(TITLE:"malaria" OR ABSTRACT:"malaria" OR KW:"malaria" OR MESH:"malaria" OR ',
    'TITLE:"plasmodium" OR ABSTRACT:"plasmodium") AND ',
    '(PUB_TYPE:"Randomized Controlled Trial" OR (SRC:"PPR" AND TITLE:randomi*)) AND ',
    '(FIRST_PDATE:[%s TO %s])'),
    start_date, end_date)
  params <- list(query = query, format = "json", resultType = "core",
                 pageSize = 100, cursorMark = "*")
  out <- list()
  seen_cursor <- character(0)
  pages <- 0L
  repeat {
    data <- http_get_json(.EPMC_BASE, params)
    for (res in data$resultList$result %||% list()) {
      out[[length(out) + 1L]] <- .parse_epmc(res)
    }
    cursor <- data$nextCursorMark %||% NULL
    pages <- pages + 1L
    if (is.null(cursor) || cursor %in% seen_cursor) break
    if (pages >= 500L) {  # ~50k records; warn rather than silently truncate
      message("[europepmc] page cap (500) reached; results may be truncated - narrow the window")
      break
    }
    seen_cursor <- c(seen_cursor, cursor)
    params$cursorMark <- cursor
  }
  out
}

.parse_epmc <- function(res) {
  pmid <- as.character(res$pmid %||% "")
  doi <- as.character(res$doi %||% "")
  src <- as.character(res$source %||% "")
  ext_id <- as.character(res$id %||% "")

  journal <- res$journalInfo$journal$title %||%
    res$bookOrReportDetails$publisher %||% ""

  ptypes <- tolower(unlist(res$pubTypeList$pubType %||% list()))
  is_rct <- any(grepl("randomi[sz]ed controlled trial", ptypes))
  is_preprint <- identical(src, "PPR")
  is_cochrane <- grepl("cochrane database syst rev", tolower(journal))

  url <- if (nzchar(doi)) {
    sprintf("https://doi.org/%s", doi)
  } else if (nzchar(pmid)) {
    sprintf("https://europepmc.org/article/MED/%s", pmid)
  } else {
    sprintf("https://europepmc.org/article/%s/%s", src, ext_id)
  }

  status <- if (is_preprint) "preprint" else if (is_cochrane) "cochrane review" else "published"
  design <- if (is_rct) "randomized controlled trial" else ""

  record_type <- if (is_preprint) "preprint" else if (is_cochrane) "review" else "trial"
  rec <- new_record(
    title = res$title %||% "", source = "europepmc",
    source_id = if (nzchar(pmid)) pmid else ext_id, url = url,
    publication_date = as.character(res$firstPublicationDate %||% ""),
    design = design, status = status, record_type = record_type,
    abstract = res$abstractText %||% "", journal = journal
  )
  rec$.doi <- doi
  rec
}

File: legacy_python/src/fetch_europepmc.py
"""Europe PMC REST search.

One robust endpoint that also folds in medRxiv preprints and Cochrane review
metadata. resultType=core returns abstracts. We page with cursorMark.
"""
from . import _http
from .record import make_record

BASE = "https://www.ebi.ac.uk/europepmc/webservices/rest/search"


def fetch(cfg, start_date, end_date):
    query = (
        'malaria AND (randomized OR randomised) AND trial '
        f'AND (FIRST_PDATE:[{start_date} TO {end_date}])'
    )
    params = {
        "query": query, "format": "json", "resultType": "core",
        "pageSize": 100, "cursorMark": "*",
    }
    out, seen_cursor, pages = [], set(), 0
    while pages < 50:
        data = _http.get(BASE, params=params).json()
        for res in data.get("resultList", {}).get("result", []):
            out.append(_parse(res))
        cursor = data.get("nextCursorMark")
        pages += 1
        if not cursor or cursor in seen_cursor:
            break
        seen_cursor.add(cursor)
        params["cursorMark"] = cursor
    return out


def _parse(res):
    pmid = res.get("pmid", "")
    doi = res.get("doi", "")
    src = res.get("source", "")
    ext_id = res.get("id", "")
    journal = res.get("journalInfo", {}).get("journal", {}).get("title", "") \
        or res.get("bookOrReportDetails", {}).get("publisher", "")
    is_preprint = src == "PPR"
    is_cochrane = "cochrane database syst rev" in (journal or "").lower()

    if doi:
        url = f"https://doi.org/{doi}"
    elif pmid:
        url = f"https://europepmc.org/article/MED/{pmid}"
    else:
        url = f"https://europepmc.org/article/{src}/{ext_id}"

    status = "preprint" if is_preprint else ("cochrane review" if is_cochrane else "published")

    return make_record(
        title=res.get("title", ""), source="europepmc",
        source_id=pmid or ext_id, url=url,
        date=res.get("firstPublicationDate", ""),
        design="", status=status,
        abstract=res.get("abstractText", ""),
    ) | {"_doi": doi, "_journal": journal}
